src/openvic-simulation/utility/Constructors.hpp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <type_traits>
+
+template<typename T>
+constexpr T value_or_move(T&& value) {
+	return std::is_rvalue_reference<T>::value ? std::move(value) : value;
```

Comment:
This won't work, even if you have a function returning a `&&` reference, the result of that needs to be moved again when initialising the variable. I was thinking to use a separate function for working out if the constructor argument should be an r-value reference or not, so that you wouldn't have to include the argument types in the macro arguments. In this case, the constructor variables should be initialised using:
```c++
ARGUMENT_NAME { \
	std::is_rvalue_reference_v<ARGUMENT_TYPE> ? std::move(ARGUMENT_NAME) : ARGUMENT_NAME \
}
```
This way the argument is actually moved into the variable, instead of being moved into an intermediate variable returned by `value_or_move`, which is then used to copy-construct the variable instead of move-constructing it. This should fix the compilation errors you were having because of  `ProductionType const&`.